lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -544,7 +445,7 @@ getTransitiveUserInBlock(Operation *baseOp, scf::ForOp &forOp) {
 }
```

Comment:
Regarding the logic above:

```
  // For now, this needs to be aligned with loadOpsToIndirectionLevelAndUse
  // since only those uses are in the schedule. Once we move all scheduling to
  // happen before lowering, any direct use will have a stage assigned.
```

should this be updated now?

Reply:
Tell me if it doesn't make sense, but seems to me that since all the ops have stages assigned, we should be looking only at the direct uses of the load?

Reply:
Although, this may break the logic of `if (use->hasTrait<OpTrait::DotLike>()) {`... I feel almost like we should decide if load is pipelined based on direct uses, and use transitive uses only to check for a dot op downstream.

Reply:
Yeah I forgot about this comment. I tried
```
-  dfs(baseOp, baseOp, false /*anyOp*/);
-  // For now, this needs to be aligned with loadOpsToIndirectionLevelAndUse
-  // since only those uses are in the schedule. Once we move all scheduling to
-  // happen before lowering, any direct use will have a stage assigned.
-  if (loopHasAttribute) {
-    seen.clear();
-    dfs(baseOp, baseOp, true /*anyOp*/);
-  }
+  dfs(baseOp, baseOp, loopHasAttribute);
```
but the lit test fails. Looking into it.

Reply:
Looking at the test case of indirect_bmm_scalar, it has a loop attribute, and this chain:
load --> convert_layout --> dot
In this case, we want to get the dot use to assign shared encoding based on it, if there is no dot in the chain, we can get the first use.

If there are multiple direct users for load:
  If there is no loop attribute, the old logic prior to refactoring will find the dots/loads along the use-chain for each edge.
  If there is a loop attribute, along edge use-edge, it will try to find the load/dot in the chain, if nothing is found, it will get the first user along the edge.
After getting the users, it goes through the users. Once a shared encoding can be found for one of the user, it will stop iterating the rest of the users.

This code kind of mimics that:
```
  dfs(baseOp, baseOp, false /*anyOp*/);
  if (loopHasAttribute) {
    seen.clear();
    dfs(baseOp, baseOp, true /*anyOp*/);
  }
```

This code doesn't do anything:
```
      // If that still didn't work, bail on pipelining this load.
      if (!loadInfo.sharedEncoding)
        continue;
```
Since it is at the end of the for loop that goes through users of the loadOp, the continue doesn't do anything. I will remove it.

Reply:
Thinking about this more, it feels wrong that we rely on the `num_stages` attribute of the loop here to figure out if we should be looking at direct or transitional users here. `num_stages` should be an information for the scheduling pass and at this point everything we should need to know should be stored in stage/cluster annotations of the ops.
For example, what if the scheduling pass generated a code like this (for whatever reason, it should be a valid IR):

```
for # no `num_stages`
  a = load # stage 0
  b = arith(a) # stage 0
  c = dot(b) # stage2
```
Because there is no `num_stages` the code will look to the dot to assign the layout, but it will also determine that the load is pipelined, while it is not. I am not saying that scheduling pass will generate something like this today, but we should be generic enough to support this case.

Reply:
I see. I was mostly trying to match with the current behavior. In general, looking at the direct user to check if a load is pipelined makes sense. We should verify load will not have multiple direct users with different stages?
In terms of assigning shared encoding, to match the current behavior, should we look at dot user first, if there is no dot user, look at the direct user?

Reply:
> I was mostly trying to match with the current behavior.
Well, that's fair. We can clean this up as a follow up. I started to look into possible ways of cleaning up scheduling and SWP further. For multiple users of a load I guess we would need to make a decision during scheduling phase